### ISSUE TITLE
Warn when `await` keyword might be missing

### ISSUE THREAD

**Issue by jgehrcke** [eslint/eslint#13567]
Picking up 
https://github.com/babel/babel-eslint/issues/593
and
https://github.com/eslint/eslint/issues/9787

Some of the meanest kinds of bugs we are running into are race conditions happening at runtime because of accidentally missing `await` statements.

Others run in this, too. Quoting @aymericbouzy:

> My real pain point is the following: I have a large codebase, and when my code runs, I encounter bugs that are difficult to reproduce, unpredictable results ... simply because I forgot to await a promise. Here is a simple example where not awaiting a promise leads to bad results:

(https://github.com/eslint/eslint/issues/9787#issuecomment-367264702)

Now, I think we already _know_ that ESLint cannot reliably detect these errors. I think that @ljharb is right with [this statement](https://github.com/eslint/eslint/issues/9787#issuecomment-367388897):

> This is what a type system and/or tests are for; this just isn’t a lintable problem.

But in practice I often times find myself in a situation where the type system doesn't catch the problem either or where the effort to make it detect the problem would simply be quite large.

Pragmatically, a simple warning when an `await` keyword _might_ be missing would do wonders.

Then I can still decide if that's the case or not.

What do you think?

**Comment by anikethsaha** [comment]
Thanks for the issue. 

Is this proposal different from what we have in [require-await](https://eslint.org/docs/rules/require-await#disallow-async-functions-which-have-no-await-expression-require-await) ?

**Comment by jgehrcke** [comment]
> Is this proposal different from what we have in require-await ?

Yes! `require-await` warns "about async functions which have no await expression". In other words: when there are zero `await` keywords. But when there is at least one then the rule is satisfied. However, that does not mean that other `await`s are not missing in the same function body :-).

**Comment by anikethsaha** [comment]
Can you give some example/code snippet like how the proposed change will work additionally with `require-await` ?

Thanks.

**Comment by jgehrcke** [comment]
> Can you give some example/code snippet like how the proposed change will work additionally with require-await?

That's a bit tough for me knowing so little about ESLint and the TypeScript ecosystem in general. Who could help us here make a reasonable and sensible proposal? Maybe the best approach would even be to extend/change the `require-await` rule? The name "require await" already sounds like the right thing (I mean, semantically).

**Comment by anikethsaha** [comment]
Actually I am not getting your proposal and how it is different from `require-await` . Some examples/snippet would have been helpful. 

Maybe I am missing something here.

**Comment by jgehrcke** [comment]
>  I am not getting your proposal and how it is different from require-await

As I wrote above: `require-await` is satisfied when it sees at least _one_ (1) `await` in a function body. However, that very function body might _still_ lack `await`s!

Let me indeed show that in an example.


```javascript
function sleep(seconds) {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000.0));
}

async function get1() {
  await sleep(1);
  return 1;
}

async function get2(input) {
  await sleep(1);
  return input;
}

async function main() {
  const result1 = get1();
  const result2 = get2(result1);
  console.log(`done: ${result2}`);
}

main();
```

The correct outcome is `done: 1` on the terminal.

Let's run this:

```
$ node example.js
done: [object Promise]
```

That is unexpected outcome, because `example.js` lacks two `await` keywords, both in `main()`.

What does ESLint with `require-await` say?

```
$ ./node_modules/eslint/bin/eslint.js . --ext .js

/home/jp/dev/eslintawaitexample/example.js
  15:1  error  Async function 'main' has no 'await' expression  require-await

✖ 1 problem (1 error, 0 warnings)
```

Ok, let's add one `await` in line 16:
```
const result1 = await get1();
```

What does ESLint with `require-await` say?

```
$ ./node_modules/eslint/bin/eslint.js . --ext .js
$ echo $?
0
```
It passes.

Is the code correct? No:

```
$ node example.js
done: [object Promise]
```

Let's add the remaining `await` keyword in line 17:
```
const result2 = await get2(result1);
```

Now the the outcome is expected:
```
$ node example.js
done: 1
```

-----------

That is, it would be highly valuable to get a warning about line 17 in this snippet:
```javascript
function sleep(seconds) {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000.0));
}

async function get1() {
  await sleep(1);
  return 1;
}

async function get2(input) {
  await sleep(1);
  return input;
}

async function main() {
  const result1 = await get1();
  const result2 = get2(result1);
  console.log(`done: ${result2}`);
}

main();
```

**Comment by anikethsaha** [comment]
Thanks a lot for the clear explanation. This is an interesting proposal but I think eslint cant be able to solve this. Because this highly needs the meta info about the functions and those can be filled using typescript. 
So I think this might be a good feature for `typescript-eslint`'s [require-await](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/require-await.md) but that would still be  a complex one to implement for example for operations coming from different modules in those cases I don't think it would be possible to support this (may be it can be done but that would be a complex one.)

So in eslint, we can't do that as eslint doest support typescript from the core.

**Comment by kaicataldo** [comment]
To clarify, it sounds like you're suggesting a rule that would warn on all function calls found in an `async` function body that are not `await`ed. Is that right?

If so, I don't think this should be a rule in core. I also don't think this is a particularly helpful rule in general, because I've never encountered a codebase where all function calls are asynchronous. This means that there would be a lot of false positives and negatives, and would most likely produce a lot of useless warnings. In my experience, this leads to people ignoring the warning altogether.

I'm a bit confused as to why a type system isn't helping you here. In my experience with TypeScript, at least, these kinds of errors have been caught.

**Comment by ljharb** [comment]
Without a type system, that’s just not possible in the most common case, which is when the others async function is in another file.

**Comment by jgehrcke** [comment]
Thanks for the feedback and thinking this through, everyone!

> Without a type system, that’s just not possible in the most common case, which is when the other async function is in another file.

Yes @ljharb -- thanks for explicitly mentioning the "is in another file" scenario. I know little about ESLint and of course if it operates on files individually then this is a severe constraint. Also, I hope you have seen that I quoted your conclusion about the type system already in my opening post! :-).


> To clarify, it sounds like you're suggesting a rule that would warn on all function calls found in an async function body that are not awaited. Is that right?

@kaicataldo Well, that's not quite what I was proposing. But if you're saying that this would be the only way via which this could be implemented in ESLint  -- then I of course agree very much that the feedback would be way too noisy!

> I'm a bit confused as to why a type system isn't helping you here. In my experience with TypeScript, at least, these kinds of errors have been caught.

Okay, I think that by now we all agree that this cannot be meaningfully warned about in ESLint, and that it has to be addressed with the type system.

I am still rather new to TypeScript, too! I am probably looking for a way to declare a type that includes "not a `Promise`" but is otherwise pretty generic.

Say, you have
```typescript
type ResultType = Dict<any> | undefined;
```
with
```typescript
export declare interface Dict<T = any> {
  [key: string]: T;
}
```
and the async function
```typescript
async function doSomething(...): Promise<ResultType> {
    ...
}
```
then
```typescript
const res: ResultType = this.doSomething();
```

will not yield a tsc error because a Promise in fact fits `Dict<any>`, i.e. the compiler does not notice when I accidentally store the promise to the result in `res`, instead of the actual result.

I just tried to explicitly exclude `Promise<any>` via `type ResultType = Exclude<Dict<any> | undefined, Promise<any>>` but that compiles just fine without the missing `await`.

For the record, I have now found these resources that seem highly related, but didn't find a neat solution yet:
- https://stackoverflow.com/questions/48944552/typescript-how-to-unwrap-remove-promise-from-a-type/48945362
- https://stackoverflow.com/questions/57951850/is-there-not-promise-type-in-typescipt
- "Add the 'awaited' type operator", https://github.com/microsoft/TypeScript/pull/35998

**Comment by ljharb** [comment]
If you typed it as `Dict<unknown>` you might get the type error you're expecting?
